File: cracking_the_coding_interview/10-5.cpp
#include <iostream>
#include <vector>
using namespace std;

// 24分

// 空の文字列はtargetにならないという前提で書いている
int searchWord(const vector<string>& words, int left, int right, string target){
    if(left > right || words.empty()){
        return -1;
    }

    int mid = (left + right) / 2;
    int diff = 0;
    bool isLeft = true;
    int check = mid;
    while(diff * 2 < right + left){
        if(mid - diff >= left && words[mid - diff] != ""){
            check = mid - diff;
            break;
        }
        if(mid + diff < right && words[mid + diff] != ""){
            check = mid + diff;
            isLeft = false;
            break;
        }
        diff++;
    }

    int ans;
    if(words[check] == target){
        return check;
    }
    if(words[check] > target){
        ans = searchWord(words, left, mid - diff, target);
    }else if(isLeft){
        ans = searchWord(words, mid + diff, right, target);
    }else{
        ans = searchWord(words, mid + diff + 1, right, target);
    }

    return ans;
}

int main(void){
    vector<string> words = {"at", "", "", "", "ball", "", "", "car", "", "", "dad", "", ""};
    cout << searchWord(words, 0, words.size(), "ball") << endl;
    cout << searchWord(words, 0, words.size(), "cat") << endl;
    cout << searchWord(vector<string>(), 0, 0, "ball") << endl;
    return 0;
}

File: cracking_the_coding_interview/10-5.py
import unittest

def find_index(words, target):
    return search(words, target, 0, len(words))

def search(words, target, left, right):
    if left >= right:
        return -1
    
    mid = (left + right) // 2
    
    start_m = mid
    while mid < right - 1 and words[mid] == "":
        mid += 1

    if words[mid] == target:
        return mid
    
    if words[mid] == "":
        return search(words, target, left, start_m)
    elif words[mid] > target:
        return search(words, target, left, start_m)
    
    return search(words, target, mid + 1, right)

class Test(unittest.TestCase):
    def test_simple(self):
        words = ["at", "", "", "", "ball", "", "", "car", "", "", "dad", "", ""]
        self.assertEqual(find_index(words, "at"), 0)
        self.assertEqual(find_index(words, "ball"), 4)
        self.assertEqual(find_index(words, "car"), 7)
        self.assertEqual(find_index(words, "dad"), 10)
        self.assertEqual(find_index(words, "bake"), -1)

if __name__ == '__main__':
    unittest.main()